rfcs/convergence/array-slots.md

```diff
@@ -0,0 +1,165 @@
+RFC: Array slots
+
+---
+
+@smhigley
+
+## Summary
+
+This RFC proposes modifying the Slots API to accept arrays of slot props to support components like Combobox or List, where the most ergonomic authoring experience is passing in an array of data rather than unique JSX children.
```

Comment:
This is how we started in [Semantic UI React](https://react.semantic-ui.com/) and we decided to use the same API in Northstar, we call it [shorthand collections](https://fluentsite.z22.web.core.windows.net/0.57.0/shorthand-props#use-with-shorthand-collections). When we started there were no other way to pass props from parent to children components:
- [React Context](https://reactjs.org/docs/context.html) was added in React 16.3
- [React Hooks](https://reactjs.org/docs/hooks-intro.html) in 16.8
- while Semantic UI React started with React 15 🙄

From our POV this API worked well as it's really simple:

```tsx
// 🦁 wow, so easy!
<Dropdown items={["Robert Tolbert", "Wanda Howard", "Tim Deboer"]} />
```

However, usage in complex applications that are our main audience is different: data comes from different sources, lists are huge and there strict performance requirements. Below is a summary of feedback that we received from Northstar customers.

### Keys are hard

First of all, customers should [use keys](https://reactjs.org/docs/lists-and-keys.html) (which is not usually obvious in examples):

```tsx
<>
  // 🟥 if we will use these strings a keys, we will get a React warning about items with duplicate keys
  <Dropdown items={["Robert", "Robert"]} />
  // ✅ will not produce warnings
  <Dropdown
    items={[
      { key: "robert-k", content: "Robert" },
      { key: "robert-t", content: "Robert" }
    ]}
  />
</>
```

There is no easy option how generate keys for customers in automated way, they **should** pass it. Check a comment from @levithomason in https://github.com/Semantic-Org/Semantic-UI-React/issues/2418#issuecomment-356689264, it's a good summary from library POV.

### Performance issues (memoization)

_"My items are really expensive to render, how I memoize them?"_

Items can really expensive to re-render especially in components like `Dropdown`, `Menu`, `List` & etc. A simple `Avatar` from Teams is below, it's even without status 😱

![image](https://user-images.githubusercontent.com/14183168/128504692-b652aac6-86c2-48b3-9c48-303c3f2431db.png)

One of the frequent asks that we had was about memoization, _we have an entry in our [FAQ](https://fluentsite.z22.web.core.windows.net/0.57.0/performance)_. Developers want to implement their features and expect that library will work intuitively for them. But it does not seem to be true from an example below:

```tsx
const MemoizedItem = React.memo(Menu.Item, (prevProps, nextProps) => {
  return prevProps.active === nextProps.active;
});

const renderBlocker = (Component, props) => <MemoizedItem {...props} />;

function App() {
  return (
    <>
      {/* 👎 Even with `renderItem()` it will not be better */}
      <Menu
        items={[
          { content: "One", key: "one", children: renderBlocker },
          { content: "Two", key: "two", children: renderBlocker }
        ]}
      />

      {/* 👍 This looks like in React docs */}
      <Menu>
        <MemoizedItem>One</MemoizedItem>
        <MemoizedItem>Two</MemoizedItem>
      </Menu>
    </>
  );
}
```

### Performance issues (redundant rendering)

Another feedback that we received is related to a fact that customers **should** compute their items in advance for menus (or dropdowns, etc.) even before they are _visible_.

It could be cheap to get data for the items itself, but each item _can_ contain a lot of additional info required to render: user's state, description, etc. For example:

```tsx
<>
  <Dropdown
    items={[
      {
        // 👇 "key" & "content" are required for rendering
        key: "robert-t",
        content: "Robert Tolbert",
        // 👇 "image" & "description" could be fetched later, but we don't allow this with "items" 😕
        image: "/avatar/RobertTolbert.jpg",
        description: "Software Engineer"
      }
    ]}
  />

  <Dropdown>
    {/* 👇 only minimal data to render "Dropdown" */}
    {[{ key: "robert-t", content: "Robert Tolbert" }].map(item => (
      // 👍 Data can be fetched lazily inside custom component
      <DropdownItemThatFetchesDetails {...item} />
    ))}
  </Dropdown>
</>
```

A short demo of this behavior is available on [this CodeSandbox](https://codesandbox.io/s/shorthands-vs-jsx-n6u0l):

- for v0 we should call `getUserStatus()` **before** a menu is opened
- for v9 we call `getUserStatus()` **only** when a menu is opened

### Performance issues (root of update is moved to the parent)

Another issue that we met is related to a fact that a parent component is a root of rendering. It's highly coupled to a previous section. See an example below:

```tsx
<>
  <Dropdown
    // 👇 "online" is a property of an item, if we will change it, Dropdown should be re-rendered *anyway*
    // 💥 Other items will be also re-rendered by default, to avoid this we will need to use memoization
    items={[{ key: "robert-t", content: "Robert Tolbert", status: "online" }]}
  />

  <Dropdown>
    {[{ key: "robert-t", content: "Robert Tolbert" }].map(item => (
      // 😅 As "status" is computed *inside* an item, only it will be updated
      <DropdownItemWithStatus {...item} />
    ))}
  </Dropdown>
</>
```

Demo of this behavior [on CodeSandbox](https://codesandbox.io/s/tender-knuth-jxjl1), check output in console:

```
// with "items"
render:DropdownA
render:DropdownItemA Jane
render:DropdownItemA John
// with JSX
render:DropdownItemBJane
```

### Bundle size penalty

If I don't use `DropdownItem` from Fluent, it will be bundled anyway with `Dropdown` as it calls it internally:

```tsx
<>
  <Dropdown
    // 👇 somewhere internally we need to render "DropdownItem", it's a direct dependency
    //    and will be bundled anyway
    items={[{ key: "robert-t", content: "Robert Tolbert", status: "online" }]}
  />

  <Dropdown>
    {/* 👇 Can be absolutely custom and used custom styles without penalties */}
    <CustomItem />
  </Dropdown>
</>
```

### Virtualization

Components like `Dropdown` (`Combobox`) usually have filtering (client search) on client-side this means that input items are different from output items:

![image](https://user-images.githubusercontent.com/14183168/128510973-d7d1fdc7-d7ca-4b74-b9ac-72646001aacb.png)
![image](https://user-images.githubusercontent.com/14183168/128511002-6479b473-c44f-404e-8344-0d3c58819c27.png)

This means that virtualization component should be put inside `Dropdown` to know about its state, so we will need to add more `props` only for it:

```tsx
<>
  <Dropdown
    items={items}
    renderItem={item => <MemoizedItem {...item} />}
    renderItems={(
      jsxItems /* items from "renderItem()" */,
      rawItems /* complex lists can require item objects */
    ) => (
      <List height={150} itemCount={items.length} itemSize={35} width={300}>
        ​{items}​
      </List>
    )}
  >
    ​{/* 💡 this should be inside */}
  </Dropdown>
  {/* VS. */}
  <Dropdown>
    ​
    <List height={150} itemCount={items.length} itemSize={35} width={300}>
      ​{MemoizedItem}​
    </List>
  </Dropdown>
</>
```

It's one of reasons why we refactored `Tree` component in Northstar to hooks, it was to complex and not enough customisable.

BTW, a I don't think that we should include virtualization into components by default:
- our customers has own solutions
- it will bundle size penalty for components

---

Libraries that were designed after React hooks appeared also use JSX children in the most cases:

- https://evergreen.segment.com/components/select
- https://chakra-ui.com/docs/form/select
- https://reach.tech/combobox
- https://material-ui.com/components/selects/
- https://react-spectrum.adobe.com/react-spectrum/ComboBox.html (supports both)

This happened mainly because of React Context & hooks, these APIs simplified sharing data & handlers between parent and children components. We also use [`use-context-selector`](https://github.com/dai-shi/use-context-selector) to avoid redundant updates in child components, there is a good showcase for it in [Northstar docs](https://fluentsite.z22.web.core.windows.net/0.57.0/components/list/definition?showCode=false&showRtl=false&showTransparent=false&showVariables=false#usage-memo) (only items with updated state are re-rendered). 

---

For components like `Combobox` I suggest to keep going with JSX items. We will meet edge cases like `Tree` component where items are nested and it will be impossible to use JSX items, but I am considering this as an exception rather than a rule.

Reply:
Thanks for the detailed comment! The issues with performance and memoization are great to bring up. I'm curious about the best way to approach some of the issues I ran into when implementing Combobox with children, especially now that the Descendants API has been removed.

Essentially, the biggest issue is that the parent control (or a parent context) needs to be aware of the number of child options present, as well as their key/id values so that the active item can be controlled. The parent is the one that receives the keyboard events, and the one that needs to set the current item's `id` in an attribute, so it needs the ability to get the previous/next/first/last/etc child's `id` in response to user interaction.

Theoretically this could work with `React.children`, as long as it's a flat list. The issue there is something like this:
```js
<Combobox>
  <Group name="fruits">
    <Option>apple</Option>
    <Option>banana</Option>
  </Group>
  <Group name="vegetables">
    <Option>onion</Option>
    <Option>zucchini</Option>
  </Group>
</Combobox>
```

Potentially we could do something like create a context that gives each option the ability to register/deregister itself by id, though it would be difficult to ensure we have the ids in the correct order. There's maybe also something we can do with `React.Children` in both Combobox _and_ Group, so Combobox's context ends up with the correct set of current options. No matter what we do, the parent is going to need to update any time the descendants array updates, just based on how activedescendant works. There's likely a clever solution to this somewhere that I haven't been able to think of yet 😊.

For memoization, since we're dealing with slot arrays, I also wonder if authoring like this might work:

```js
const MemoOption = React.memo(Option);

<Combobox options=[
  <MemoOption props={...etc} />
  <MemoOption props={...etc} />
  <MemoOption props={...etc} />
] />
```

I need to test it more, but it seems like that might still allow delayed updates and allowing the author to update one child without affecting the others. I'm definitely not confident on that approach yet, I'd definitely need to test it more (or you might know offhand if there's a reason it wouldn't fly). It definitely makes managing Combobox's state easier though 😅.

Reply:
Virtualization is also a really interesting case, since I think even with options as children, the virtualization provider would likely need to live inside `Combobox` as a slot, since Combobox itself also needs to know what the total number of options is vs. the rendered number.

Reply:
Chatted @smhigley and @behowell about this RFC.  After pondering a bit, I think having a slot that takes an array of literals, props objects, or Components is what I would expect as consistent with singular slots in other components.  A common scenario will be to pass in the array of props and be able to have a Component that is given each of the those props as a template.  What if there was an itemAs property that behaved like as (literal tag name, component, or render function) that would be used for each item?  This extends the as metaphor from single slots to a slot with an array.

Reply:
> the virtualization provider would likely need to live inside Combobox as a slot, since Combobox itself also needs to know what the total number of options is vs. the rendered number.

With the v9 hooks/render func composition architecture, I would have expected users to simply create their own component variants and wrap children with the virtualization provider